pkg/istio-agent/README.md

```diff
@@ -0,0 +1,82 @@
+# Istio Agent
+
+This document describes the internal architecture of Istio agent.
+
+## High Level overview
+
+![High Level overview](docs/overview.svg)
+
+At a high level, the Istio agent acts as an intermediate proxy between Istiod and Envoy. This is done
+at two levels. For distributing workload certificates, Envoy will send [SDS](https://www.envoyproxy.io/docs/envoy/latest/configuration/security/secret)
+requests to the agent, causing the agent to submit a CSR to the configured CA (generally Istiod). For other configuration,
+Envoy will send [ADS](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/operations/dynamic_configuration#aggregated-xds-ads)
+requests to the agent, which will be forwarded to the configured discovery server (general Istiod).
+
+## CA Flow
+
+![CA Flow](docs/ca.svg)
+
+A single SDS request from Envoy goes through a few different layers in istio-agent.
+
+1. First, the request is handled by the SDS server. This is mostly just an intermediate translation layer exposing
+   the `SecretManager` to Envoy, without much business logic. For each resource requested by Envoy, the SDS server
+   will call `SecretManager.GenerateSecret(resourceName)`
+1. When `GenerateSecret` is called, the `SecretManager` is expected to return a new certificate. This can occur in a few ways.
+   1. The most common method (pictured above) is to sign a new certificate by calling the configured CA. Typically, this is Istiod.
+   1. If the certificate is not yet expired, the `SecretManager` also can return a cache response. In practice, this would
+       only happen in cases where Envoy were to re-request a resource, which is fairly rare.
+   1. `SecretManager` can also read certificates from files. When this is configured, no CA client is used.
+1. The `caClient` will be configured to use either JWT or mTLS authentication. For JWT authentication, gRPC's `PerRPCCredentials`
+   is configured with a `TokenProvider` which handles the logic of adding the proper JWT to each request. mTLS is configured
+   by a tls.Config that points to files on disk.
+
+It should be noted there is a circular dependency with mTLS authentication; in order to fetch a certificate we need
+a certificate. This can be handled in various ways:
+* `GenerateSecret` may additionally write any signed certificates to disk, with `OUTPUT_CERTS` configured.
```

Comment:
Worth clarifying the 2 options:
- we can load existing certs created by a 3p tool, for example mounted Secrets or CSI - in this case we will NOT call Istiod to get certs, just use the file.
- for VM we may also use the on-disk cert to request a fresh cert - in this case OUTPUT_CERTS allows us to write down the new cert in the same place. There are 2 cases, where the provisioning cert is long-lived cert and we exchange it for a short-lived cert - in which case OUTPUT_CERTS will not be set or will point to a different dir, or cases where the original cert is also short lived and we need to rotate.

Finally, even if if JWT are the source - OUTPUT_CERTS is used to save the cert for other apps ( like prometheus, gRPC).